docs/grammar/Solidity.g4

```diff
@@ -0,0 +1,248 @@
+        (modifierInvocation | Payable | visibility)*
+        body=block;
+
+stateMutability: Pure | View | Payable;
+overrideSpecifier: Override (LParen overrides+=userDefinedTypeName (Comma overrides+=userDefinedTypeName)* RParen)?;
+functionDefinition:
+	Function name=(identifier | Fallback | Receive)
+	LParen (arguments=parameterList)? RParen
+	(modifierInvocation | stateMutability | visibility | Virtual | overrideSpecifier)*
+	(Returns LParen returnParameters=parameterList RParen)?
+	(Semicolon | body=block);
+
+modifierDefinition:
+    Modifier name=identifier
+    (LParen (arguments=parameterList)? RParen)?
+    (Virtual | overrideSpecifier)*
+    (Semicolon | body=block);
+
+fallbackReceiveFunctionDefinition:
+	kind=(Fallback | Receive) LParen RParen
+	(visibility | stateMutability | modifierInvocation | Virtual | overrideSpecifier)*
+	(Semicolon | body=block);
+
+structMemberDefinition: type=typeName name=identifier Semicolon;
+structDefinition: Struct name=identifier LBrace members=structMemberDefinition+ RBrace;
+
+enumDefinition:	Enum name=identifier LBrace enumValues+=identifier (Comma enumValues+=identifier)* RBrace;
+
+stateVariableDeclaration:
+	type=typeName
+	(Public | Private | Internal | Constant | Virtual | overrideSpecifier | Immutable)*
+	name=identifier
+	(Assign initialValue=expression)?
+	Semicolon;
+
+eventParameter: type=typeName Indexed? name=identifier?;
+eventDefinition:
+    Event name=identifier
+    LParen (parameters+=eventParameter (Comma parameters+=eventParameter)*)? RParen
+    Anonymous?
+    Semicolon;
+
+usingDirective: Using userDefinedTypeName For (Mul | typeName) Semicolon;
+
+typeName: elementaryTypeName | functionTypeName | mappingType | userDefinedTypeName | typeName LBrack expression? RBrack;
+elementaryTypeName: Address Payable? | Bool | String | Bytes | Int | Uint | FixedBytes | Fixed | Ufixed;
+functionTypeName:
+    Function LParen (arguments=parameterList)? RParen
+    (visibility | stateMutability)*
+    (Returns LParen returnParameters=parameterList RParen)?;
+
+variableDeclaration: type=typeName location=storageLocation? name=identifier;
+storageLocation: Memory | Storage | Calldata;
+
+
+variableDeclarationList: variableDeclarations+=variableDeclaration (Comma variableDeclarations+=variableDeclaration)*;
+
+
+expression:
+	expression LBrack index=expression? RBrack # IndexAccess
+	| expression LBrack start=expression? Colon end=expression? RBrack # IndexRangeAccess
+	| expression Period (identifier | Address) # MemberAccess
+	| expression LBrace (namedArgument (Comma namedArgument)*)? RBrace # FunctionCallOptions
+	| expression functionCallArgumentList # FunctionCall
+	| Payable functionCallArgumentList # PayableConversion
+	| Type LParen typeName RParen # MetaType
+	| (Inc | Dec | Not | BitNot | Delete | Sub) expression # UnaryPrefixOperation
+	| expression (Inc | Dec) # UnarySuffixOperation
+	|<assoc=right> expression Exp expression # ExpOperation
```

Comment:
Is `|<assoc=right>` the syntax (no whitespace after the pipe) or just convention?

Reply:
I think it's only convention (which I'd be happy to break).